postgresql/src/main/scala/org/scalarelational/postgresql/PostgreSQLDatastore.scala

```diff
@@ -0,0 +1,145 @@
+package org.scalarelational.postgresql
+
+import javax.sql.DataSource
+
+import org.postgresql.ds.PGSimpleDataSource
+import org.powerscala.log.{Level, Logging}
+import org.powerscala.property.Property
+import org.scalarelational.column.ColumnPropertyContainer
+import org.scalarelational.column.property.{AutoIncrement, Polymorphic, Unique}
+import org.scalarelational.datatype._
+import org.scalarelational.instruction.ddl.CreateColumn
+import org.scalarelational.model._
+import org.scalarelational.op.{Condition, RegexCondition}
+import org.scalarelational.result.ResultSetIterator
+
+import scala.collection.mutable.ListBuffer
+
+/**
+ * @author Robert Djubek <envy1988@gmail.com>
+ */
+
+sealed trait PGSsl
+case class PGUseSsl(sslfactory: Option[String] = None, sslfactoryarg: Option[String] = None ) extends PGSsl
+case class PGNoSsl() extends PGSsl
+
+case class PGConfig(host: String,
+                            schema: String,
+                            user: String,
+                            password: String,
+                            port: Int = 5432,
+                            useSsl: PGSsl = PGNoSsl()
+                           )
+
+abstract class PostgreSQLDatastore private() extends SQLDatastore with Logging with SQLLogging {
+  protected def this(pgConfig: PGConfig) = {
+    this()
+    sqlLogLevel := Level.Warn
+    config := pgConfig
+  }
+
+  override def supportsMerge = false
+
+  protected def this(dataSource: DataSource) = {
+    this()
+    dataSourceProperty := dataSource
+  }
+
+  Class.forName("org.postgresql.Driver")
+
+  val config = Property[PGConfig]()
+
+  init()
+
+  protected def init() = {
+    config.change.on {
+      case evt => updateDataSource() // Update the data source if the mode changes
+    }
+  }
+
+  def updateDataSource() = {
+    dispose() // Make sure to shut down the previous DataSource if possible
+
+    val source: PGSimpleDataSource = new PGSimpleDataSource()
+    source.setPortNumber(config().port)
+    source.setServerName(config().host)
+    source.setDatabaseName(config().schema)
+    source.setUser(config().user)
+    source.setPassword(config().password)
+    config().useSsl match {
+      case PGUseSsl(sslfactory, sslfactoryarg) =>
+        source.setSsl(true)
+        sslfactory.foreach {f => source.setSslfactory(f)}
+        sslfactoryarg.foreach {f => source.setSslFactoryArg(f)}
+      case PGNoSsl() =>
+    }
+    dataSourceProperty := source
+  }
+
+  override protected def columnPropertiesSQL(container: ColumnPropertyContainer): List[String] = {
+    val b = ListBuffer.empty[String]
+    if (!container.isOptional && !container.has(Polymorphic)) {
+      b.append("NOT NULL")
+    }
+    if (container.has(Unique)) {
+      b.append("UNIQUE")
+    }
+    b.toList
+  }
+
+  override protected def columnSQLType(create: CreateColumn[_]) =
```

Comment:
We should come up with a better approach. This potentially doesn't cover a lot of cases. For example, ``VARCHAR_IGNORECASE`` which is used by ``StringDataType`` doesn't exist in PostgreSQL. I would suggest to create an abstraction for SQL data types, so that we can better change the behaviour of data types depending on the driver.